File: Editorial/0164_Maximum_Gap/JavaScript/3_Buckets_and_The_Pigeonhole_Principle.js
class Bucket {
    constructor() {
        this.used = false;
        this.minval = Number.MAX_SAFE_INTEGER;
        this.maxval = Number.MIN_SAFE_INTEGER;
    }
}

var maximumGap = function (nums) {
    if (!nums || nums.length < 2) return 0;

    let mini = Math.min(...nums),
        maxi = Math.max(...nums);

    let bucketSize = Math.max(1, Math.floor((maxi - mini) / (nums.length - 1))); // bucket size or capacity
    let bucketNum = Math.floor((maxi - mini) / bucketSize) + 1; // number of buckets
    let buckets = Array.from({ length: bucketNum }, () => new Bucket());

    for (let num of nums) {
        let bucketIdx = Math.floor((num - mini) / bucketSize); // locating correct bucket
        buckets[bucketIdx].used = true;
        buckets[bucketIdx].minval = Math.min(num, buckets[bucketIdx].minval);
        buckets[bucketIdx].maxval = Math.max(num, buckets[bucketIdx].maxval);
    }

    let prevBucketMax = mini,
        maxGap = 0;
    for (let bucket of buckets) {
        if (!bucket.used) continue;

        maxGap = Math.max(maxGap, bucket.minval - prevBucketMax);
        prevBucketMax = bucket.maxval;
    }

    return maxGap;
};

File: Editorial/164_Maximum_Gap/C/3_Buckets_and_The_Pigeonhole_Principle.c
#define MIN(i, j) ((i) < (j) ? (i) : (j))
#define MAX(i, j) ((i) > (j) ? (i) : (j))

typedef struct {
    int used;
    int minval;
    int maxval;
} Bucket;

int maximumGap(int* nums, int numsSize) {
    if (!nums || numsSize < 2) return 0;

    int mini = nums[0], maxi = nums[0];
    for (int i = 0; i < numsSize; i++) {
        mini = MIN(mini, nums[i]);
        maxi = MAX(maxi, nums[i]);
    }

    int bucketSize =
        MAX(1, (maxi - mini) / (numsSize - 1));      // bucket size or capacity
    int bucketNum = (maxi - mini) / bucketSize + 1;  // number of buckets
    Bucket* buckets = (Bucket*)malloc(sizeof(Bucket) * bucketNum);
    for (int i = 0; i < bucketNum; i++) {
        buckets[i].used = 0;
        buckets[i].minval = INT_MAX;
        buckets[i].maxval = INT_MIN;
    }

    for (int i = 0; i < numsSize; i++) {
        int bucketIdx =
            (nums[i] - mini) / bucketSize;  // locating correct bucket
        buckets[bucketIdx].used = 1;
        buckets[bucketIdx].minval = MIN(nums[i], buckets[bucketIdx].minval);
        buckets[bucketIdx].maxval = MAX(nums[i], buckets[bucketIdx].maxval);
    }

    int prevBucketMax = mini, maxGap = 0;
    for (int i = 0; i < bucketNum; i++) {
        if (!buckets[i].used) continue;

        maxGap = MAX(maxGap, buckets[i].minval - prevBucketMax);
        prevBucketMax = buckets[i].maxval;
    }

    free(buckets);
    return maxGap;
}
